data-structures/src/data/structure/linkedlist/LinkedList.java

```diff
@@ -0,0 +1,239 @@
+package data.structure.linkedlist;
+import static  java.util.Objects.*;
+
+public class LinkedList<E> {
+    private class Node {
+        E data;
+        Node next;
+        Node previous;
+
+        Node(final E data) {
+            this.data = data;
+            this.next = null;
+            this.previous = null;
+        }
+    }
+
+    private Node head;
+    private Node tail;
+    private int size;
+
+    public LinkedList() {
+        this.head = null;
+        this.tail = null;
+        size = 0;
+    }
+
+    public void add(E element) {
+        requireNonNull(element, "The key must not be null!");
+
+        if (isNull(head)) {
+            head = new Node(element);
+            tail = head;
+            size++;
+
+        } else {
+            tail.next = new Node(element);
+            final Node temp = tail;
+            tail = tail.next;
+            tail.previous = temp;

```

Comment:
Whoa, now this is complex! Have you thought of something like this?

~~~~Java
final Node newTail = new Node(element);
newTail.previous = tail;
tail.next = newTail;
tail = newTail;
~~~~

Both versions are 4 lines long, however, there is a great semantical difference: naming and data flow. While the original code uses a completely unrelated name, `temp`, the above snippet is much more explicit with `newTail`. Also, the flow is easier to follow: we set the references and then replace the tail.